crowbar_framework/lib/crowbar/chef/upload.rb

```diff
@@ -0,0 +1,284 @@
+          uploader = ::Chef::CookbookUploader.new(
+            cookbooks_for_upload,
+            force: false,
+            concurrency: cookbooks_for_upload.count
+          )
+          return false unless uploader.upload_cookbooks
+
+          true
+        end
+
+        def data_bags
+          loader = ::Chef::Knife::Core::ObjectLoader.new(::Chef::DataBagItem, logger)
+          data_bags = loader.find_all_object_dirs(chef_data_bags_path) || []
+          data_bags.each do |data_bag|
+            data_bag_items = loader.find_all_objects(chef_data_bags_path.join(data_bag))
+            data_bag_item_paths = normalize_data_bag_item_paths(data_bag_items) || []
+            data_bag_item_paths.each do |data_bag_item_path|
+              # Workaround for a strange chef behavior
+              relative_path = chef_data_bags_path.relative_path_from(Pathname.new(Dir.pwd))
+              data_bag_item = loader.load_from(relative_path, data_bag, data_bag_item_path)
+              bag = ::Chef::DataBagItem.new
+              bag.data_bag(data_bag)
+              bag.raw_data = data_bag_item
+              logger.info("Uploading data_bag #{data_bag_item_path}")
+
+              return false unless bag.save
+            end
+          end
+          true
+        end
+
+        def roles
+          loader = ::Chef::Knife::Core::ObjectLoader.new(::Chef::Role, logger)
+          chef_roles_path.each_child do |file|
+            next unless file.file?
+            role = loader.load_from("roles", file)
+            logger.info("Uploading role #{file.basename}")
+
+            return false unless role.save
+          end
+          true
+        end
+
+        protected
+
+        def logger
+          return Rails.logger unless caller.grep(/rake/).present?
+          ::Logger.new(STDOUT)
+        end
+
+        def chef_path
+          Rails.root.join("..", "chef")
+        end
+
+        def chef_cookbooks_path
+          chef_path.join("cookbooks")
+        end
+
+        def chef_data_bags_path
+          chef_path.join("data_bags")
+        end
+
+        def chef_roles_path
+          chef_path.join("roles")
+        end
+
+        def version_constraint(version)
+          ::Chef::VersionConstraint.new(version)
+        end
+
+        def cookbook_loader
+          @cookbook_loader ||= ::Chef::CookbookLoader.new(chef_cookbooks_path)
+        end
+
+        def cookbooks_to_upload
+          cookbook_loader.load_cookbooks
+        end
+
+        def check_dependencies(cookbook)
+          missing_dependencies = cookbook.metadata.dependencies.reject do |cookbook_name, version|
+            check_server_side_cookbooks(cookbook_name, version) ||
+              check_uploading_cookbooks(cookbook_name, version)
+          end
+
+          unless missing_dependencies.empty?
+            missing_cookbooks = missing_dependencies.map do |cookbook_name, version|
+              "\"#{cookbook_name}\" version \"#{version}\""
```

Comment:
Because I want to print these quotes whiteout escaping the string would be broken.